README


This archive contains the distribution Config-MVP,
version 2.200008:

  multivalue-property package-oriented configuration

This software is copyright (c) 2014 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.019.



Changes
Revision history for Config-MVP

2.200008  2014-07-02 19:24:59-04:00 America/New_York
        - fix a busted link
        - be more consistent in exception-throwing on missing packages
        - fix a dumb test that failed in a git checkout

2.200007  2014-01-29 21:04:25-05:00 America/New_York
        - throw an exception when a bundle reports an impossible situation,
          like { value => [] }; https://github.com/rjbs/Config-MVP/issues/5

2.200006  2013-09-17 08:58:53 Asia/Tokyo
        - fix repository data

2.200005  2013-09-17 08:56:36 Asia/Tokyo
        - removed use of deprecated Class::MOP::load_class (thanks, Karen
          Etheridge)

2.200004  2013-08-27 12:09:19 America/New_York
        - pass the arguments to missing_plugin in the right order

2.200003  2013-01-30 21:35:34 America/New_York
        - doc fixes (Thanks, George Hartzell!)

2.200002  2012-03-15 21:16:16 America/New_York
        - documentation fixes (Thanks, Gregor Herrmann of Debian!)

        - depend on at least version 0.17 of Class::Load, to work around a bug
          with perl5.8.8 and Module::Runtime 0.012. (Karen Etheridge)

2.200001  2011-02-11 11:13:20 America/New_York
        - when throwing "package not installed", add a "package" attribute to
          Error

2.200000  2011-02-08 11:05:12 America/New_York
        - add Config::MVP::Error for common exceptions

        - allow class loading and class-not-found handling to be subclassed
          (thanks, Kent Fredric!)

2.101650  2010-06-14 20:22:42 America/New_York
        - require Moose 0.91 for native trait support

2.101540  2010-06-03 00:28:57 America/New_York
        - improve if_none to return from the proper scope (i.e., now it works)

2.101480  2010-05-28 22:59:18 America/New_York
        - add an if_none callback to Reader::Finder; for now, private

2.101460  2010-05-26 22:29:43 America/New_York
          [BREAKS BACKWARD COMPATIBILITY]

        - I apologize for breaking backcompat twice in a week.  I do not
          expect this to happen again.  -- rjbs

        - Config::MVP::Reader is now a base class rather than a role.
          This significantly simplifies the use of Config::MVP readers
          and makes things easier to document.

1.101451  2010-05-25 22:38:06 America/New_York
        - always finalize sections added by bundles

1.101450  2010-05-25 18:13:19 America/New_York
          [BREAKS BACKWARD COMPATIBILITY]

        - The Reader interface has been substantially overhauled.

        - Readers no longer have persistent Assembler, but now must be handed
          one (or will generate one as needed)

        - the arguments for Reader->read_config have changed

        - classes implementing Reader are required to provide a
          read_into_assembler method, now

        - Reader::Findable has been broken into Findable and
          Findable::ByExtension

0.101410  2010-05-21 22:09:48 America/New_York
          Sections, Sequences, and Assemblers  can now be finalized

          the Assembler finalizes things as appropriate

          Sections can refer to their containing Sequence

          Sequences can refer to their constructing Assembler

          bundles now can (and should) provide payload as an arrayref

          Assembler->begin_section can be given a reference to a package name
          to skip rewriting

          the replace_bundle_with_contents method is used by
          Assembler::WithBundles to convert bundle sections into contents

0.100780  2010-03-19 22:16:16 America/New_York
          the alias attr on Section is now lazy so that mvp_aliases is called

0.093350  2009-12-01 06:58:48 America/New_York
          added Config::MVP::Reader::Hash for use by adapter writers

0.093330  2009-11-29 17:40:08 America/New_York
          rebuild to include META.yml for sake of cpandeps scanner

0.093120  2009-11-08
          add missing prereq

0.093000  2009-10-27
          bundles for Assembler::WithBundles can now be nested

0.092990  2009-10-26
          add Config::MVP::Reader and associated packages, from Dist::Zilla

0.092360  2009-08-24
          add "begin_section" and "end_section" methods, mostly for wrapping
          with extra behavior in subclasses

0.092211  2009-08-09
          much more documentation yet again

0.092210  2009-08-09
          package interrogation (for mvp_* settings) is now done in Section
            initialization; this means you don't need to use the Assembler to
            benefit from this feature

          much more documentation



lib/Config/MVP/Section.pm
package Config::MVP::Section;
# ABSTRACT: one section of an MVP configuration sequence
$Config::MVP::Section::VERSION = '2.200007';
use Moose 0.91;

use Class::Load 0.17 ();
use Config::MVP::Error;

# =head1 DESCRIPTION
#
# For the most part, you can just consult L<Config::MVP> to understand what this
# class is and how it's used.
#
# =attr name
#
# This is the section's name.  It's a string, and it must be provided.
#
# =cut

has name => (
  is  => 'ro',
  isa => 'Str',
  required => 1
);

# =attr package
#
# This is the (Perl) package with which the section is associated.  It is
# optional.  When the section is instantiated, it will ensure that this package
# is loaded.
#
# =cut

has package => (
  is  => 'ro',
  isa => 'Str', # should be class-like string, but can't be ClassName
  required  => 0,
  predicate => 'has_package',
);

# =attr multivalue_args
#
# This attribute is an arrayref of value names that should be considered
# multivalue properties in the section.  When added to the section, they will
# always be wrapped in an arrayref, and they may be added to the section more
# than once.
#
# If this attribute is not given during construction, it will default to the
# result of calling section's package's C<mvp_multivalue_args> method.  If the
# section has no associated package or if the package doesn't provide that
# method, it default to an empty arrayref.
#
# =cut

has multivalue_args => (
  is   => 'ro',
  isa  => 'ArrayRef',
  lazy => 1,
  default => sub {
    my ($self) = @_;

    return []
      unless $self->has_package and $self->package->can('mvp_multivalue_args');

    return [ $self->package->mvp_multivalue_args ];
  },
);

# =attr aliases
#
# This attribute is a hashref of name remappings.  For example, if it contains
# this hashref:
#
#   {
#     file => 'files',
#     path => 'files',
#   }
#
# Then attempting to set either the "file" or "path" setting for the section
# would actually set the "files" setting.
#
# If this attribute is not given during construction, it will default to the
# result of calling section's package's C<mvp_aliases> method.  If the
# section has no associated package or if the package doesn't provide that
# method, it default to an empty hashref.
#
# =cut

has aliases => (
  is   => 'ro',
  isa  => 'HashRef',
  lazy => 1,
  default => sub {
    my ($self) = @_;

    return {} unless $self->has_package and $self->package->can('mvp_aliases');

    return $self->package->mvp_aliases;
  },
);

# =attr payload
#
# This is the storage into which properties are set.  It is a hashref of names
# and values.  You should probably not alter the contents of the payload, and
# should read its contents only.
#
# =cut

has payload => (
  is  => 'ro',
  isa => 'HashRef',
  init_arg => undef,
  default  => sub { {} },
);

# =attr is_finalized
#
# This attribute is true if the section has been marked finalized, which will
# prevent any new values from being added to it.  It can be set with the
# C<finalize> method.
#
# =cut

has is_finalized => (
  is  => 'ro',
  isa => 'Bool',
  traits   => [ 'Bool' ],
  init_arg => undef,
  default  => 0,
  handles  => { finalize => 'set' },
);

before finalize => sub {
  my ($self) = @_;

  Config::MVP::Error->throw("can't finalize unsequenced Config::MVP::Section")
    unless $self->sequence;
};

# =attr sequence
#
# This attributes points to the sequence into which the section has been
# assembled.  It may be unset if the section has been created but not yet placed
# in a sequence.
#
# =cut

has sequence => (
  is  => 'ro',
  isa => 'Config::MVP::Sequence',
  weak_ref  => 1,
  predicate => '_sequence_has_been_set',
  reader    => '_sequence',
  writer    => '__set_sequence',
  clearer   => '_clear_sequence',
);

sub _set_sequence {
  my ($self, $seq) = @_;

  Config::MVP::Error->throw("Config::MVP::Section cannot be resequenced")
    if $self->sequence;

  $self->__set_sequence($seq);
}

sub sequence {
  my ($self) = @_;
  return undef unless $self->_sequence_has_been_set;
  my $seq = $self->_sequence;

  Config::MVP::Error->throw("can't access section's destroyed sequence")
    unless defined $seq;

  return $seq;
}

# =method add_value
#
#   $section->add_value( $name => $value );
#
# This method sets the value for the named property to the given value.  If the
# property is a multivalue property, the new value will be pushed onto the end of
# an arrayref that will store all values for that property.
#
# Attempting to add a value for a non-multivalue property whose value was already
# added will result in an exception.
#
# =cut

sub add_value {
  my ($self, $name, $value) = @_;

  confess "can't add values to finalized section " . $self->name
    if $self->is_finalized;

  my $alias = $self->aliases->{ $name };
  $name = $alias if defined $alias;

  my $mva = $self->multivalue_args;

  if (grep { $_ eq $name } @$mva) {
    my $array = $self->payload->{$name} ||= [];
    push @$array, $value;
    return;
  }

  if (exists $self->payload->{$name}) {
    Carp::croak "multiple values given for property $name in section "
              . $self->name;
  }

  $self->payload->{$name} = $value;
}

# =method load_package
#
#   $section->load_package($package, $plugin);
#
# This method is used to ensure that the given C<$package> is loaded, and is
# called whenever a section with a package is created.  By default, it delegates
# to L<Class::Load>.  If the package can't be found, it calls the
# L<missing_package> method.  Errors in compilation are not suppressed.
#
# =cut

sub load_package {
  my ($self, $package, $plugin) = @_;

  Class::Load::load_optional_class($package)
    or $self->missing_package($package, $plugin);
}

# =method missing_package
#
#   $section->missing_package($package, $plugin);
#
# This method is called when C<load_package> encounters a package that is not
# installed.  By default, it throws an exception.
#
# =cut

sub missing_package {
  my ($self, $package, $plugin) = @_ ;

  my $class = Moose::Meta::Class->create_anon_class(
    superclasses => [ 'Config::MVP::Error' ],
    cached       => 1,
    attributes   => [
      Moose::Meta::Attribute->new(package => (
        is       => 'ro',
        required => 1,
      )),
    ],
  );

  $class->name->throw({
    ident   => 'package not installed',
    message => "$package (for plugin $plugin) does not appear to be installed",
    package => $package,
  });
}

sub _BUILD_package_settings {
  my ($self) = @_;

  return unless defined (my $pkg = $self->package);

  confess "illegal package name $pkg" unless Params::Util::_CLASS($pkg);

  $self->load_package($pkg, $self->name);

  # We call these accessors for lazy attrs to ensure they're initialized from
  # defaults if needed.  Crash early! -- rjbs, 2009-08-09
  $self->multivalue_args;
  $self->aliases;
}

sub BUILD {
  my ($self) = @_;
  $self->_BUILD_package_settings;
}

no Moose;
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Config::MVP::Section - one section of an MVP configuration sequence

=head1 VERSION

version 2.200007

=head1 DESCRIPTION

For the most part, you can just consult L<Config::MVP> to understand what this
class is and how it's used.

=head1 ATTRIBUTES

=head2 name

This is the section's name.  It's a string, and it must be provided.

=head2 package

This is the (Perl) package with which the section is associated.  It is
optional.  When the section is instantiated, it will ensure that this package
is loaded.

=head2 multivalue_args

This attribute is an arrayref of value names that should be considered
multivalue properties in the section.  When added to the section, they will
always be wrapped in an arrayref, and they may be added to the section more
than once.

If this attribute is not given during construction, it will default to the
result of calling section's package's C<mvp_multivalue_args> method.  If the
section has no associated package or if the package doesn't provide that
method, it default to an empty arrayref.

=head2 aliases

This attribute is a hashref of name remappings.  For example, if it contains
this hashref:

  {
    file => 'files',
    path => 'files',
  }

Then attempting to set either the "file" or "path" setting for the section
would actually set the "files" setting.

If this attribute is not given during construction, it will default to the
result of calling section's package's C<mvp_aliases> method.  If the
section has no associated package or if the package doesn't provide that
method, it default to an empty hashref.

=head2 payload

This is the storage into which properties are set.  It is a hashref of names
and values.  You should probably not alter the contents of the payload, and
should read its contents only.

=head2 is_finalized

This attribute is true if the section has been marked finalized, which will
prevent any new values from being added to it.  It can be set with the
C<finalize> method.

=head2 sequence

This attributes points to the sequence into which the section has been
assembled.  It may be unset if the section has been created but not yet placed
in a sequence.

=head1 METHODS

=head2 add_value

  $section->add_value( $name => $value );

This method sets the value for the named property to the given value.  If the
property is a multivalue property, the new value will be pushed onto the end of
an arrayref that will store all values for that property.

Attempting to add a value for a non-multivalue property whose value was already
added will result in an exception.

=head2 load_package

  $section->load_package($package, $plugin);

This method is used to ensure that the given C<$package> is loaded, and is
called whenever a section with a package is created.  By default, it delegates
to L<Class::Load>.  If the package can't be found, it calls the
L<missing_package> method.  Errors in compilation are not suppressed.

=head2 missing_package

  $section->missing_package($package, $plugin);

This method is called when C<load_package> encounters a package that is not
installed.  By default, it throws an exception.

=head1 AUTHOR

Ricardo Signes <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
